core/dbt/task/runnable.py

```diff
@@ -72,17 +71,20 @@ def __init__(self, args, config, manifest):
         self.node_results = []
         self.num_nodes: int = 0
         self.previous_state: Optional[PreviousState] = None
+        self.previous_defer_state: Optional[PreviousState] = None
         self.run_count: int = 0
         self.started_at: float = 0
 
-        self.set_previous_state()
-
-    def set_previous_state(self):
-        if self.args.state is not None:
+        if self.args.state:
             self.previous_state = PreviousState(
                 path=self.args.state, current_path=Path(self.config.target_path)
             )
 
+        if self.args.defer_state:
+            self.previous_defer_state = PreviousState(
```

Comment:
Am I right in understanding - this could potentially increase memory footprint by the size of the manifest?

> Seems like defer actually happens after graph selection is done. A optimization we can do later on is after the graph selection, drop the state and replace it with defer_state. But that's trickier to reason then the current one.

That's an interesting idea! It is definitely trickier to reason about, but so long as we comment/document it well, I'm not strictly opposed. I _think_ this also work with the `state_relation` piece we'd need for #7256 (which should be based on `defer_state`, rather than logical-comparison state).

Reply:
That sounds good to me. As this is a new flag if people run into issue with memory it would not be such a bad thing. Do worth calling out when handing over to shipment team.